File: 130-surrounded-regions/surrounded-regions.cpp
class Solution {
public:
    void dfs(vector<vector<char>> &board, int i, int j){
        if(i < 0 || i >= board.size() || j < 0 || j >= board[0].size() || board[i][j] == '~') return;
        board[i][j] = '~';
        if(i < board.size()-1 && board[i+1][j] == 'O') dfs(board, i+1, j);
        if(i > 0 && board[i-1][j] == 'O') dfs(board, i-1,j);
        if(j < board[0].size()-1 && board[i][j+1] == 'O') dfs(board, i, j+1);
        if(j > 0 && board[i][j-1] == 'O') dfs(board, i, j-1);
    }
    void solve(vector<vector<char>>& board) {
        for (int i=0; i<board[0].size(); i++){
            if(board[0][i] == 'O') dfs(board,0,i);
            if(board[board.size()-1][i] == 'O') dfs(board, board.size()-1,i);
        } 

        for(int i=0; i<board.size(); i++){
            if(board[i][0] == 'O') dfs(board,i,0);
            if(board[i][board[0].size()-1] == 'O') dfs(board,i,board[0].size()-1);
        }

        for(int i=0; i<board.size(); i++){
            for(int j=0; j<board[0].size(); j++){
                if(board[i][j] == '~') board[i][j] = 'O';
                else if(board[i][j] == 'O') board[i][j] = 'X';
            }
        }
    }
};

File: 130-surrounded-regions/surrounded-regions.py
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        m, n = len(board), len(board[0])

        def dfs(i,j):
            nonlocal board
            if i < 0 or i >= m or j < 0 or j >= n:
                return
            if board[i][j] == '$' or board[i][j] == 'X':
                return
            board[i][j] = '$'
            offset = [(0,1), (0,-1), (1,0), (-1,0)]
            for of in offset:
                dfs(i+of[0], j+of[1])

        # Check the edges
        for i in range(n):
            if board[0][i] == 'O':
                dfs(0,i)
            if board[m-1][i] == 'O':
                dfs(m-1,i)
        
        for i in range(m):
            if board[i][0] == 'O':
                dfs(i,0)
            if board[i][n-1] == 'O':
                dfs(i,n-1)

        for i in range(m):
            for j in range(n):
                if board[i][j] == '$':
                    board[i][j] = 'O'
                elif board[i][j] == 'O':
                    board[i][j] = 'X'
        